internal/database/external_services.go

```diff
@@ -992,15 +991,36 @@ func (e *externalServiceStore) Delete(ctx context.Context, id int64) (err error)
 	}
 	defer func() { err = tx.Done(err) }()
 
-	// We take an advisory lock here and also when syncing an external service to
-	// ensure that they can't happen at the same time.
-	lock := locker.NewWith(tx, "external_service")
-	locked, err := lock.LockInTransaction(ctx, locker.StringKey(fmt.Sprintf("%d", id)), false)
+	// Load the external service *for update* so that no sync job can be created
+	if err := tx.selectForUpdate(ctx, id); err != nil {
+		return err
+	}
+
+	// Cancel all currently running sync jobs, *outside* the transaction.
+	err = e.CancelSyncJob(ctx, ExternalServicesCancelSyncJobOptions{ExternalServiceID: id})
 	if err != nil {
-		return errors.Wrap(err, "getting advisory lock")
+		return err
 	}
-	if !locked {
-		return errors.Errorf("could not get advisory lock for service %d", id)
+
+	// Wait until all the sync jobs we just canceled are done executing to
+	// ensure that we delete all repositories and no new ones are inserted.
+	runningJobsCtx, cancel := context.WithTimeout(ctx, 45*time.Second)
+	defer cancel()
+
+	for {
+		if err := runningJobsCtx.Err(); err != nil {
+			return err
+		}
+
+		ok, err := e.hasRunningSyncJobs(runningJobsCtx, id)
```

Comment:
nit: more obvious naming

```suggestion
		runningJobsExist, err := e.hasRunningSyncJobs(runningJobsCtx, id)
```

Reply:
Haha, I actually had that and then went with `ok` in the hopes of being more idiomatic.